---
title: "class18"
author: "Nguyen"
date: "3/5/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Connect to the NCI-GDC database

The NCI-GDC (cancer genomic commons) is a major portal for cancer genomics data.

Here we use bioconductor package **GenomicDataCommons** to talk to this datavase and download data etc. all from R.

```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

Can we talk to the online database?

```{r}
status()
```


## Querying the GDC from R

The pipe operator can be useful in R and sometimes is more clear to read.
```{r}
c(1:10) %>% mean()
```


```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```
### Plot this overview data as a barplot
```{r}
x <- cases_by_project$project.project_id
barplot(x$doc_count)
```
##Data is highly skewed so we will plot the log of this data instead
```{r}
barplot(x$doc_count, names.arg = x$key, log = "y",  las = 2, col="lightblue")
```


##Variant analysis with R

Find files available PAAD project that we can
```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
head(maf.file)
```

```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)
```

```{r}
plotmafSummary(vars)
```


##Oncoplots


##Hands on Part 2

>Q1. What are the tumor specific amino acids?

```{r}
library(bio3d)

seqs <- read.fasta("lecture18_sequences.fa") 
seqs
```

```{r}
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```

## Make a "names" label for our output sequences (one per mutant)
```{r}
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```

## Sequence positions surounding each mutant site
```{r}
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```


```{r}
x <- c(1:6)
y <- c(1:4)
x %in% y
```

Find the -8 before and +8 after (i.e. the sequence segment containing all possible 9-mer sequences with a tumor specific amino-acid)

```{r}
mutant.sites -8
mutant.sites + 8
```

```{r}
seqs$ali[,start.position[1]:end.position[1]]
```