---
title: "class16: Pathway Analysis from RNA Seq"
author: "Nguyen"
date: "2/27/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import RNA-Seq counts and metadata
The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

Here we read the two input files

```{r}
# Import metadata and take a peak
colData = read.csv("GSE37704_metadata.csv", row.names=1)
head(colData)
```

Add out count data

```{r}
countData = read.csv("GSE37704_featurecounts.csv", row.names=1)
head(colData)
```

There are `r nrow(countData)` genes in this dataset and
`r ncol(countData)` experiments (samples)...

Looks like there is a funny 'length' column in our countData that we need to remove. (there should only be 6 genes, but there ar 7 columns)


```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```

A little bit more tidying work required. Lets remove the genes where we have no data (i. e. zero counts in all experiments)
(find the sums of the rows, get rid of those that have a sum of zero)

Lets take the rowSums() and exclude the rows with zero values

```{r}
countData <- countData[rowSums(countData) != 0, ]
```

Check how genes we have remaining
```{r}
nrow(countData)
```

## RUnning DESeq2

Now we get to

```{r}
library(DESeq2)

dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

Lets see our results

```{r}
res <- results(dds)
res
```

## Volcano plot
Summary volcano plot of our results to date.

```{r}
plot( res$log2FoldChange, -log(res$padj) )
abline(h=log(0.05), col="red")
abline(v=c(-2,2), col="red")
```

Lets add some color...
# Make a color vector for all genes
```{r}
mycols <- rep("gray", nrow(res) )
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"
plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```

#Adding gene annotation

Since we mapped and counted against the Ensembl annotation, our results only have information about Ensembl gene IDs. However, our pathway analysis downstream will use KEGG pathways, and genes in KEGG pathways are annotated with Entrez gene IDs. So lets add them as we did the last day.

I would install these with
`BiocManager::install("AnnotationDbi")` and
`BiocManager::install("org.Hs.eg.db")`

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)

#What is available int the org.Hs.eg.db dataset?
columns(org.Hs.eg.db)
```

Lets add extra annotation to our result object
```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                     column="SYMBOL",
                    multiVals="first")

res$entrez <- mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                     column="ENTREZID",
                    multiVals="first")
  
head(res)
```


Now we have our geneIDS in the correct format

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

First I need to install the packages I need

```{r}
# BiocManager::install( c("pathview", "gage", "gageData") )
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now run **gage** with this input vector and our KEGG genesets

```{r}
#Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

What is in our results? We can use the `attributes()` function to see what is returned here.

```{r}
attributes(keggres)
```


```{r}
# Look at the first few down (less) pathways
head(keggres$less)

```

We use the KEGG pathway ID from our top hit along with our `foldchange` vector to view out results as a more useful figure

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```


```{r}
knitr::include_graphics("hsa04110.pathview.png")
```

And one more...
```{r}
head(keggres$greater)

```


```{r}
pathview(gene.data=foldchanges, pathway.id="hsa00140")
```


```{r}
knitr::include_graphics("hsa00140.pathview.png")
```




